### PR TITLE
Fixed #21528 -- Document parent access in formfield_for_foreignkey

### PR DESCRIPTION
**Pull request by francoisfreitag**
Since this is my first commit to Django, I started with some documentation.

If it looks good to you, I'll try to implement a nicer way to retrieve parent's ID in ``ModelAdmin``.

### CODE REVIEW

**Comment by timgraham** [comment]
Regarding the first commit, I think https://github.com/django/django/pull/5872 is a better approach as I don't think it should be implied that `request` may be `None`. Regarding the third commit, I'd like to see the implementation that you come up with before we decided to say "This is a bit of a hack and should be improved in future versions." If we can fix this is an elegant fashion, then I think we might skip documenting the "hack". Thanks!

**Comment by francoisfreitag** [comment]
I would pass the parent object to `formfield_for_foreignkey` this way: https://github.com/django/django/commit/d8dd30c0bc5400f0938e5b05114320cf95947e8e.

If it seems like the way to go, I will add a deprecation warning when `parent` is present in the method kwargs (meaning that the old `formfield_for_foreignkey` definition is used).

**Comment by timgraham** [comment]
What do you think about the approach [suggested on Stackoverflow](http://stackoverflow.com/a/949290/5112) of using a custom `ModelForm` instead?

**Comment by francoisfreitag** [comment]
My use cases would be covered by the ModelForm approach.

However, beginners might find it easier to override a method in the `ModelAdmin` than to override the queryset in a ModelForm `__init__()`. They'll probably stumble upon `self.instance.pk` being `None` while it's not expected.

My preferred option would be to add the keyword argument to `formfield_for_xxx()` as it seems pretty clean (except the deprecation trick) and I can't find a reason not to do it. Another option would be to point to `ModelForm` in the `formfield_for_xxx()` documentation.
What would you prefer?

**Comment by timgraham** [comment]
I'm not set against it, but my feeling is that it's better to promote patterns that are reusable outside of the admin. Adding more and more hooks to the admin for every little thing is a nuisance. Other opinions welcome, but I interpret this as an agreement with my view: https://groups.google.com/d/msg/django-developers/WrnhmTyLHuY/ldXK3aOGgv4J.

File: docs/ref/forms/fields.txt [hunk 1]
```diff
@@ -1077,14 +1079,24 @@ objects (in the case of ``ModelMultipleChoiceField``) into the
 
 For more complex uses, you can specify ``queryset=None`` when declaring the
 form field and then populate the ``queryset`` in the form's ``__init__()``
-method::
+method:
+
+.. code-block:: python
+
+    class Country(models.Model):
```
Comment (timgraham) [inline docs/ref/forms/fields.txt]: Not sure what's wrong with the existing example from ced3e303? This seems to add more complexity in trying to understand the essence of the issue (of course, I wrote the original patch, so I'm biased). The example no longer demonstrates `queryset=None` as stated on line 1080.

File: docs/ref/contrib/admin/index.txt [hunk 2]
```diff
@@ -1615,8 +1615,21 @@ templates used by the :class:`ModelAdmin` views:
                     kwargs["queryset"] = Car.objects.filter(owner=request.user)
                 return super(MyModelAdmin, self).formfield_for_foreignkey(db_field, request, **kwargs)
 
-    This uses the ``HttpRequest`` instance to filter the ``Car`` foreign key
-    field to only display the cars owned by the ``User`` instance.
+    This uses the :class:`~django.http.HttpRequest` instance to filter the
```
Comment (timgraham) [inline docs/ref/contrib/admin/index.txt]: I guess it'd be more consistent to link every instance of HttpRequest in this document, although I'm not convinced it adds much.

File: docs/ref/forms/fields.txt [hunk 3]
```diff
@@ -1077,14 +1079,24 @@ objects (in the case of ``ModelMultipleChoiceField``) into the
 
 For more complex uses, you can specify ``queryset=None`` when declaring the
 form field and then populate the ``queryset`` in the form's ``__init__()``
-method::
+method:
+
+.. code-block:: python
```
Comment (timgraham) [inline docs/ref/forms/fields.txt]: The default highlighting should work without this unless the code block has a syntax error that causes the parsing to fail.

**Comment by timgraham** [comment]
Sorry for the delayed review. Please ensure "Patch needs improvement" is unchecked on the ticket after you update the patch so that the ticket appears in the review queue.

**Comment by timgraham** [comment]
Closing due to inactivity. Please open a new PR if you can update it.

**Pull request closed without merging**